README
Vumi Wikipedia App
==================

This is a vumi[1] app that provides access to Wikipedia over a combination of
SMS and USSD.

To install:

    virtualenv --no-site-packages ve
    source ./ve/bin/activate
    pip install -r requirements.pip
    pip install -e .


NOTE: vumi 0.3.1 (the current stable release) is too old for this app, and the
      dev branch in github has a problem with one of its deps. There are some
      workarounds for this in the requirements.pip file.

[1] http://vumi.org


debian/changelog
vumi-wikipedia (0.1~a+1-0~ppa3) precise; urgency=low

  * Initial release.

 -- Colin Alston <colin.alston@praekeltfoundation.org>  Wed, 25 Jul 2012 10:27:22 +0200


vumi_wikipedia/tests/test_wikipedia.py
"""Tests for vumi.demos.wikipedia."""

from twisted.internet.defer import inlineCallbacks

from vumi.application.tests.test_base import ApplicationTestCase

from vumi_wikipedia.wikipedia import WikipediaWorker
from vumi_wikipedia.tests.test_wikipedia_api import (
    FakeHTTPTestCaseMixin, WIKIPEDIA_RESPONSES)


class WikipediaWorkerTestCase(ApplicationTestCase, FakeHTTPTestCaseMixin):
    application_class = WikipediaWorker

    @inlineCallbacks
    def setUp(self):
        yield super(WikipediaWorkerTestCase, self).setUp()
        yield self.start_webserver(WIKIPEDIA_RESPONSES)
        self.worker = yield self.get_wikipedia_worker()
        self.wikipedia = self.worker.wikipedia

    @inlineCallbacks
    def tearDown(self):
        yield self.stop_webserver()
        yield super(WikipediaWorkerTestCase, self).tearDown()

    def get_wikipedia_worker(self, config_override=None):
        config = {
            'transport_name': self.transport_name,
            'worker_name': 'wikitest',
            'sms_transport': 'sphex_sms',
            'api_url': self.url,
            }
        if config_override is not None:
            config.update(config_override)
        return self.get_application(config)

    def assert_config_knob(self, attr, orig, new):
        self.assertEqual(orig, getattr(self.worker, attr))
        self.assertEqual(new, getattr(self.knobbly_worker, attr))

    @inlineCallbacks
    def search_for_content(self, search, result=1, section=1):
        yield self.dispatch(self.mkmsg_in(None))  # Start session.
        yield self.dispatch(self.mkmsg_in(search))  # Search keyword.
        yield self.dispatch(self.mkmsg_in(str(result)))  # Select result.
        yield self.dispatch(self.mkmsg_in(str(section)))  # Select section.

    def test_make_options(self):
        self.assertEqual((2, "1. foo\n2. bar"),
                         self.worker.make_options(['foo', 'bar']))

    @inlineCallbacks
    def test_happy_flow(self):
        yield self.dispatch(self.mkmsg_in(None))
        self.assertEqual('What would you like to search Wikipedia for?',
                         self.get_dispatched_messages()[-1]['content'])

        yield self.dispatch(self.mkmsg_in('cthulhu'))
        self.assertEqual('\n'.join([
                    u'1. Cthulhu',
                    u'2. Call of Cthulhu (role-playing game)',
                    u'3. Cthulhu (2007 film)',
                    u'4. Cthulhu (2000 film)',
                    u'5. Cthulhu Mythos',
                    u'6. The Call of Cthulhu',
                    ]),
                         self.get_dispatched_messages()[-1]['content'])

        yield self.dispatch(self.mkmsg_in('1'))
        self.assertEqual('\n'.join([
                    u'1. Cthulhu',
                    u'2. History',
                    u'3. Geography',
                    u'4. Mountains of Madness',
                    u'5. Lulz',
                    ]),
                         self.get_dispatched_messages()[-1]['content'])

        yield self.dispatch(self.mkmsg_in('2'))
        self.assertEqual(
            u'The first half of the principal manuscript told a very peculiar '
            u'tale. It appears that on 1 March 1925, a thin, dark young man of'
            u'...\n(Full content sent by SMS.)',
            self.get_dispatched_messages()[-1]['content'])
        [sms_msg] = self._amqp.get_messages('vumi', 'sphex_sms.outbound')
        self.assertEqual(u'The first half of the principal manuscript told a '
            u'very peculiar tale. It appears that on 1 March 1925, a thin, '
            u'dark young man of neurotic and excited aspect had...',
            sms_msg['content'])

    @inlineCallbacks
    def test_search_no_results(self):
        yield self.dispatch(self.mkmsg_in(None))
        self.assertEqual('What would you like to search Wikipedia for?',
                         self.get_dispatched_messages()[-1]['content'])

        yield self.dispatch(self.mkmsg_in('ncdkiuagdqpowebjkcs'))
        self.assertEqual(
            'Sorry, no Wikipedia results for ncdkiuagdqpowebjkcs',
            self.get_dispatched_messages()[-1]['content'])

    @inlineCallbacks
    def test_search_error(self):
        yield self.dispatch(self.mkmsg_in(None))
        self.assertEqual('What would you like to search Wikipedia for?',
                         self.get_dispatched_messages()[-1]['content'])

        yield self.dispatch(self.mkmsg_in('.'))
        self.assertEqual(
            'Sorry, there was an error processing your request. Please try '
            'again later.', self.get_dispatched_messages()[-1]['content'])
        self.flushLoggedErrors()

    @inlineCallbacks
    def test_config_knobs(self):
        self.knobbly_worker = yield self.get_wikipedia_worker({
                'api_url': 'https://localhost:1337/',
                'accept_gzip': True,
                'user_agent': 'Bob Howard',
                'max_ussd_session_length': 200,
                'content_cache_time': 1800,
                'max_ussd_content_length': 180,
                'max_ussd_unicode_length': 80,
                'max_sms_content_length': 300,
                'max_sms_unicode_length': 130,
                })

        self.assert_config_knob('api_url', self.url, 'https://localhost:1337/')
        self.assert_config_knob('accept_gzip', None, True)
        self.assert_config_knob('user_agent', None, 'Bob Howard')
        self.assert_config_knob('max_ussd_session_length', 180, 200)
        self.assert_config_knob('content_cache_time', 3600, 1800)
        self.assert_config_knob('max_ussd_content_length', 160, 180)
        self.assert_config_knob('max_ussd_unicode_length', 70, 80)
        self.assert_config_knob('max_sms_content_length', 160, 300)
        self.assert_config_knob('max_sms_unicode_length', 70, 130)


vumi_wikipedia/tests/test_text_manglers.py
# -*- coding=utf-8 -*-

from twisted.trial.unittest import TestCase

from vumi_wikipedia.text_manglers import (mangle_text, convert_unicode,
    normalize_whitespace, is_unicode, truncate_sms, truncate_sms_with_postfix)


class TextManglersTestCase(TestCase):

    def test_mangle_text(self):
        rev = lambda t: ''.join(reversed(t))
        upp = lambda t: t.upper()
        add = lambda t: ''.join([t, 'd'])
        self.assertEqual(u'abc', mangle_text(u'abc'))
        self.assertEqual(u'cba', mangle_text(u'abc', [rev]))
        self.assertEqual(u'abc', mangle_text(u'abc', [rev, rev]))
        self.assertEqual(u'CBA', mangle_text(u'abc', [rev, upp]))
        self.assertEqual(u'dCBA', mangle_text(u'abc', [upp, add, rev]))
        self.assertEqual(u'CBAD', mangle_text(u'abc', [rev, add, upp]))

    def test_convert_unicode(self):
        self.assertEqual(u'a-b c', convert_unicode(u'a\u2013b\xa0c'))

    def test_normalize_whitespace(self):
        self.assertEqual(u'a b c', normalize_whitespace(u'\ta  b\n c\r'))

    def test_is_unicode(self):
        self.assertFalse(is_unicode(u'@foo^bar!'))
        self.assertTrue(is_unicode(u'foobar \n превед'))

    def test_truncate_sms(self):
        self.assertEquals(u'', truncate_sms(u''))
        self.assertEquals(u'foo...', truncate_sms(u'foo bar', 6, 3))
        self.assertEquals(u'Спасибо Пукину...',
            truncate_sms(u'Спасибо Пукину за это', 30, 15))

    def test_truncate_sms_with_postfix(self):
        self.assertEquals(u'foo bar... (for more madness, program in Python)',
            truncate_sms_with_postfix(u'foo bar baz',
            u' (for more madness, program in Python)', 46, 23))
        self.assertEquals(u'хрень какая-то... (testetstest)',
            truncate_sms_with_postfix(u'хрень какая-то нахреначилася',
            u' (testetstest)', 60, 30))
        self.assertEquals(u'foo bar... (превед)',
            truncate_sms_with_postfix(u'foo bar baz',
            u' (превед)', 32, 16))


vumi_wikipedia/tests/test_wikipedia_api.py
import json
from functools import wraps
from pkg_resources import resource_stream

from twisted.internet import reactor
from twisted.internet.defer import inlineCallbacks
from twisted.internet.protocol import Protocol, Factory
from twisted.trial.unittest import TestCase

from vumi_wikipedia.wikipedia_api import WikipediaAPI, ArticleExtract, APIError


class SectionMarkerCreator(object):
    def __getitem__(self, key):
        return u'\ufffd\ufffd%s\ufffd\ufffd' % (key,)


def make_extract(text):
    return ArticleExtract(text % SectionMarkerCreator())


class ArticleExtractTestCase(TestCase):
    def test_one_section(self):
        ae = make_extract(u'foo\nbar')
        self.assertEqual([u'foo\nbar'], ae.get_section_texts())
        self.assertEqual([], ae.get_section_titles())
        self.assertEqual(
            [{'title': None, 'level': None, 'text': u'foo\nbar'}], ae.sections)
        self.assertEqual([{'title': None, 'level': None, 'text': u'foo\nbar'}],
            ae.get_top_level_sections())

    def test_multiple_sections(self):
        ae = make_extract(u'foo\n\n\n%(2)s bar \nbaz\n%(2)squux\n\n\nlol')
        self.assertEqual([u'foo', u'baz', u'lol'], ae.get_section_texts())
        self.assertEqual([u'bar', u'quux'], ae.get_section_titles())

    def test_nested_sections(self):
        ae = make_extract(u'%(2)sfoo\n%(3)s bar \ntext')
        self.assertEqual([u'', u'', u'text'], ae.get_section_texts())
        self.assertEqual([u'foo', u'bar'], ae.get_section_titles())
        self.assertEqual([{'level': None, 'text': u'', 'title': None},
            {'level': 2, 'text': u'', 'title': u'foo'}
            ], ae.get_top_level_sections())

    def test_empty_input(self):
        ae = ArticleExtract(u'')
        self.assertEqual([u''], ae.get_section_texts())
        self.assertEqual([], ae.get_section_titles())
        self.assertEqual([{'title': None, 'level': None, 'text': u''}],
            ae.sections)


WIKIPEDIA_RESPONSES = json.load(
    resource_stream(__name__, 'wikipedia_responses.json'))


class FakeHTTP(Protocol):
    def dataReceived(self, data):
        request_line, body = self.parse_request(data)
        response = self.handle_request(request_line, body)
        self.transport.write(response.encode('utf-8'))
        self.transport.loseConnection()

    def parse_request(self, data):
        headers, _, body = data.partition('\r\n\r\n')
        headers = headers.splitlines()
        request_line = headers.pop(0).rsplit(' ', 1)[0]
        self.assert_user_agent(headers)
        return request_line, body

    def assert_user_agent(self, headers):
        expected_user_agent = getattr(
            self.factory.testcase, 'expected_user_agent', None)
        if expected_user_agent is not None:
            [user_agent] = [h.split(': ', 1)[1] for h in headers
                            if h.lower().startswith('user-agent')]
            self.factory.testcase.assertEqual(expected_user_agent, user_agent)

    def build_response(self, response_data):
        lines = ["HTTP/1.1 %s" % (response_data['response_code'],)]
        lines.extend(['', json.dumps(response_data['response_body'])])
        return '\r\n'.join(lines)

    def handle_request(self, request_line, body):
        response_data = self.factory.response_data.get(request_line)
        if not response_data:
            self.factory.testcase.fail(
                "Unexpected request: %s" % (request_line,))
        self.factory.testcase.assertEqual(response_data["request_body"], body)
        return self.build_response(response_data)


class FakeHTTPTestCaseMixin(object):
    @inlineCallbacks
    def start_webserver(self, response_data):
        factory = Factory()
        factory.protocol = FakeHTTP
        factory.response_data = response_data
        factory.testcase = self
        self.webserver = yield reactor.listenTCP(0, factory)
        addr = self.webserver.getHost()
        self.url = "http://%s:%s/" % (addr.host, addr.port)

    def stop_webserver(self):
        return self.webserver.loseConnection()


def debug_api_call(func):
    @wraps(func)
    def wrapped_test(self):
        self.wikipedia.PRINT_DEBUG = True
        self.wikipedia.url = self.wikipedia.URL
        return func(self)
    return wrapped_test


class WikipediaAPITestCase(TestCase, FakeHTTPTestCaseMixin):
    timeout = 10

    @inlineCallbacks
    def setUp(self):
        yield self.start_webserver(WIKIPEDIA_RESPONSES)
        self.wikipedia = WikipediaAPI(self.url, False)

    def tearDown(self):
        return self.stop_webserver()

    def assert_api_result(self, api_result_d, expected):
        return api_result_d.addCallback(self.assertEqual, expected)

    @inlineCallbacks
    def test_search_success(self):
        yield self.assert_api_result(
            self.wikipedia.search('wikipedia', limit=3),
            [u'Wikipedia', u'Wikip\xe9dia', u'Main Page'])
        # And again with a different request and result limit
        yield self.assert_api_result(
            self.wikipedia.search('vumi', limit=2),
            [u'Arambagh Utsab', u'Vulpia microstachys'])

    def test_search_error(self):
        return self.assertFailure(self.wikipedia.search('.'), APIError)

    def test_search_no_results(self):
        return self.assert_api_result(
            self.wikipedia.search('ncdkiuagdqpowebjkcs'), [])

    def test_get_extract(self):
        def assert_extract(extract):
            self.assertEqual(5, len(extract.sections))

        return self.wikipedia.get_extract('Cthulhu').addCallback(
            assert_extract)

    @inlineCallbacks
    def test_user_agent(self):
        self.expected_user_agent = self.wikipedia.USER_AGENT
        yield self.wikipedia.get_extract('Cthulhu')
        self.wikipedia = WikipediaAPI(self.url, False, 'Bob Howard')
        self.expected_user_agent = 'Bob Howard'
        yield self.wikipedia.get_extract('Cthulhu')


build-packages.py
#!/usr/bin/python

import os
import sys
import time


buildNumber = os.getenv('BUILD_NUMBER')
if not buildNumber:
    print "No build number"
    sys.exit(1)

sourceVersion = "0.1~a+%s" % (buildNumber,)
baseDir = os.getcwd()
tarball = 'vumi-wikipedia_%s.orig.tar.gz' % (sourceVersion,)


def configBuild(platform, platformver, dhmake):
    if os.path.exists('./debian/rules'):
        os.remove('./debian/rules')

    rules = open('./debian/rules', 'wt')
    rules.write("#!/usr/bin/make -f\n\n%%:\n"
                "\t\tdh \"$@\" --with %s\noverride_dh_clean:\n"
                "\t\tdh_clean\n\t\trm -rf *.egg-info\n" % dhmake)
    rules.close()

    os.chmod('./debian/rules', 0755)

    dch = """vumi-wikipedia (%s-0~ppa%s) %s; urgency=low

  * Initial release.

 -- Colin Alston <colin.alston@praekeltfoundation.org>  %s\n""" % (
        sourceVersion,
        platformver,
        platform,
        time.strftime('%a, %d %b %Y %H:%M:%S %z')
    )
    changelog = open('./debian/changelog', 'wt')
    changelog.write(dch)
    changelog.close()


def prepareBuild():
    os.chdir(baseDir)
    os.system('mkdir -p dist/vumi-wikipedia')
    os.system('git archive HEAD -o dist/%s' % (tarball,))


def runBuild(platform, platformver, dhmake):
    os.chdir(os.path.join(baseDir, 'dist/vumi-wikipedia'))
    os.system('tar xzf ../%s' % (tarball,))
    configBuild(platform, platformver, dhmake)
    os.system('debuild -S')


prepareBuild()
runBuild('lucid', 1, 'quilt,python-central')
runBuild('precise', 3, 'quilt,python2')


vumi_wikipedia/text_manglers.py
# -*- test-case-name: vumi_wikipedia.tests.test_text_manglers -*-

import unicodedata
import re


def mangle_text(text, manglers=()):
    for mangler in manglers:
        text = mangler(text)
    return text


def unicode_ord(name):
    return ord(unicodedata.lookup(name))


UNICODE_CONVERSION_MAPPING = {
    unicode_ord('EN DASH'): u'-',
    unicode_ord('EM DASH'): u'-',
    }


def convert_unicode(text):
    """Convert unicode characters to ASCII, where possible.

    First step is to normalize whatever we've been given using the
    compatibility decomposition and canonical recomposition (NFKC). Thereafter,
    we manually replace some extra characters.

    NOTE: This does not strip out all non-ASCII characters. Just some of them.
    """

    text = unicodedata.normalize('NFKC', text)
    return text.translate(UNICODE_CONVERSION_MAPPING)


def normalize_whitespace(text):
    """Replace each whitespace sequence with a single space.
    """
    return ' '.join(text.strip().split())


UNICODE_REGEX = re.compile(u'[\u0080-\uffff]')


def is_unicode(string):
    return UNICODE_REGEX.search(string) != None


def truncate_sms(string, ascii_limit=160, unicode_limit=70, ellipsis=u'...'):
    """Smart string truncation
    """
    result = u''
    for word in string.split(' '):
        longer_string = (result + ' ' + word).strip()
        if (((is_unicode(longer_string) and len(longer_string) > unicode_limit)
            or (len(longer_string) > ascii_limit))):
                return result + ellipsis
        result = longer_string
    return result


def truncate_sms_with_postfix(string, postfix, ascii_limit=160,
    unicode_limit=70, ellipsis=u'...'):
        length = len(postfix)
        if is_unicode(postfix):
            ascii_limit = unicode_limit
        return truncate_sms(string, ascii_limit - length,
            unicode_limit - length, ellipsis) + postfix


vumi_wikipedia/wikipedia_api.py
# -*- test-case-name: vumi_wikipedia.tests.test_wikipedia_api -*-

import re
import json
from urllib import urlencode

from twisted.internet.defer import inlineCallbacks, returnValue

from vumi.utils import http_request_full


def either(*args):
    for arg in args:
        if arg is not None:
            return arg
    return None


ARTICLE_SPLITTER = re.compile(u'\ufffd\ufffd(?=\\d)')
ARTICLE_SECTION = re.compile(
    u'^(\\d)\ufffd\ufffd\\s*([^\\n]+?)\\s*(?:|\\n+(.*))$', re.DOTALL)


class APIError(Exception):
    """
    Exception thrown by Wikipedia API.
    """


class ArticleExtract(object):
    """
    Class representing an article extract
    """

    def __init__(self, data):
        if isinstance(data, unicode):
            self._init_from_string(data)
        else:
            self.sections = data

    def _init_from_string(self, string):
        self.sections = []
        for section in ARTICLE_SPLITTER.split(string):
            section = section.strip()
            if len(self.sections) > 0:
                m = ARTICLE_SECTION.match(section)
                level, title, text = m.groups()
                level = int(level)
                if text is None:
                    text = u''
            else:
                title, text, level = (None, section, None)
            self.sections.append(
                {'title': title, 'level': level, 'text': text})

    def get_section_titles(self):
        return [section['title'] for section in self.sections[1:]]

    def get_section_texts(self):
        return [section['text'] for section in self.sections]

    def get_top_level_sections(self):
        level = 10
        result = []
        for section in self.sections:
            if section['level'] is None or section['level'] <= level:
                result.append(section)
            if section['level'] is not None:
                level = min(level, section['level'])
        return result


class WikipediaAPI(object):
    """
    Small Wikipedia API client library.

    :param str url: URL of the API to query. (Defaults to Wikipedia's API.)
    :param bool gzip: `True` to ask for gzip encoding, `False` otherwise.
    """

    URL = 'http://en.wikipedia.org/w/api.php'

    # The MediaWiki API docs request that clients use gzip encoding to reduce
    # network traffic. However, Twisted only supports this easily from 11.1.
    GZIP = False

    USER_AGENT = 'Vumi HTTP Request'

    PRINT_DEBUG = False

    def __init__(self, url=None, gzip=None, user_agent=None):
        self.url = either(url, self.URL)
        self.gzip = either(gzip, self.GZIP)
        self.user_agent = either(user_agent, self.USER_AGENT)

    @inlineCallbacks
    def _make_call(self, params):
        params.setdefault('format', 'json')
        url = '%s?%s' % (self.url, urlencode(params))
        headers = {'User-Agent': self.user_agent}
        if self.gzip:
            headers['Accept-Encoding'] = 'gzip'
        if self.PRINT_DEBUG:
            print "\n=====\n\n%s /?%s\n" % ('GET', url.split('?', 1)[1])
        response = yield http_request_full(url, '', headers, method='GET')
        if self.PRINT_DEBUG:
            print response.delivered_body
            print "\n====="
        returnValue(json.loads(response.delivered_body))

    @inlineCallbacks
    def search(self, query, limit=9):
        """
        Perform a query and returns a list of results matching the query.

        :param unicode query: Search terms.
        :param int limit: Maximum number of results to return. (Default 9)

        :returns: `list` of article titles matching search terms.
        """
        response = yield self._make_call({
                'action': 'query',
                'list': 'search',
                'srsearch': query.encode('utf-8'),
                'srlimit': str(limit),
                })
        if 'query' not in response:
            raise APIError(response)
        results = [r['title']
                   for r in response['query'].get('search', {})]
        returnValue(results)

    @inlineCallbacks
    def get_extract(self, page_name):
        """
        Return the content of a section of a page.

        :param unicode page_name: The name of the page to query.

        :returns: :class:`ArticleExtract` containing the article data.
        """
        response = yield self._make_call({
                'action': 'query',
                'prop': 'extracts',
                'explaintext': '',
                'exsectionformat': 'raw',
                'titles': page_name.encode('utf-8'),
                'redirects': '1',
                })
        if 'query' not in response:
            raise APIError(response)
        _id, page = response['query']['pages'].popitem()
        returnValue(ArticleExtract(page['extract']))


vumi_wikipedia/wikipedia.py
# -*- test-case-name: vumi_wikipedia.tests.test_wikipedia -*-

import json

from twisted.internet.defer import inlineCallbacks, returnValue
from twisted.python import log
from vumi.application import ApplicationWorker
from vumi.persist.txredis_manager import TxRedisManager
from vumi.components.session import SessionManager

from vumi_wikipedia.wikipedia_api import WikipediaAPI, ArticleExtract
from vumi_wikipedia.text_manglers import (
    normalize_whitespace, truncate_sms, truncate_sms_with_postfix)


class WikipediaUSSDFlow(object):
    def __init__(self, worker, session):
        self.session = session


def mkmenu(options, prefix, start=1):
    return prefix + '\n'.join(
        ['%s. %s' % (idx, opt) for idx, opt in enumerate(options, start)])


class WikipediaWorker(ApplicationWorker):
    """Look up Wikipedia content over USSD, deliver over USSD/SMS.

    Config parameters
    -----------------

    sms_transport : str, optional
        If set, this specifies a different transport for sending SMS replies.
        Otherwise the same transport will be used for both USSD and SMS.

    override_sms_address : str, optional
        If set, this overrides the `to_addr` for SMS replies. This is useful
        for demos where a fake USSD transport is being used but real SMS
        replies are desired.

    api_url : str, optional
        Alternate API URL to use. This can be any MediaWiki deployment,
        although certain assumptions are made about the structure of articels
        that may not be valid outside of Wikipedia.

    accept_gzip : bool, optional
        If `True`, the HTTP client will request gzipped responses. This is
        generally beneficial, although it requires Twisted 11.1 or later.

    user_agent : str, optional
        Override `User-Agent` header on API requests.

    max_ussd_session_length : int, optional
        Lifetime of USSD session in seconds. Defaults to 3 minutes.

    content_cache_time : int, optional
        Lifetime of cached article content in seconds. Defaults to 1 hour.

    max_ussd_content_length : int, optional
        Maximum character length of ASCII USSD content. Defaults to 160.

    max_ussd_unicode_length : int, optional
        Maximum character length of unicode USSD content. Defaults to 70.

    max_sms_content_length : int, optional
        Maximum character length of ASCII SMS content. Defaults to 160.

    max_sms_unicode_length : int, optional
        Maximum character length of unicode SMS content. Defaults to 70.
    """

    MAX_USSD_SESSION_LENGTH = 3 * 60
    CONTENT_CACHE_TIME = 3600

    MAX_USSD_CONTENT_LENGTH = 160
    MAX_USSD_UNICODE_LENGTH = 70
    MAX_SMS_CONTENT_LENGTH = 160
    MAX_SMS_UNICODE_LENGTH = 70

    def _opt_config(self, name):
        return self.config.get(name, None)

    def validate_config(self):
        self.sms_transport = self._opt_config('sms_transport')
        self.override_sms_address = self._opt_config('override_sms_address')
        self.api_url = self._opt_config('api_url')
        self.accept_gzip = self._opt_config('accept_gzip')
        self.user_agent = self._opt_config('user_agent')
        self.max_ussd_session_length = self.config.get(
            'max_ussd_session_length', self.MAX_USSD_SESSION_LENGTH)
        self.content_cache_time = self.config.get(
            'content_cache_time', self.CONTENT_CACHE_TIME)
        self.max_ussd_content_length = self.config.get(
            'max_ussd_content_length', self.MAX_USSD_CONTENT_LENGTH)
        self.max_ussd_unicode_length = self.config.get(
            'max_ussd_unicode_length', self.MAX_USSD_UNICODE_LENGTH)
        self.max_sms_content_length = self.config.get(
            'max_sms_content_length', self.MAX_SMS_CONTENT_LENGTH)
        self.max_sms_unicode_length = self.config.get(
            'max_sms_unicode_length', self.MAX_SMS_UNICODE_LENGTH)

    @inlineCallbacks
    def setup_application(self):
        redis = yield TxRedisManager.from_config(
            self.config.get('redis_manager', {}))
        redis = redis.sub_manager(self.config['worker_name'])

        self.extract_redis = redis.sub_manager('extracts')

        self.session_manager = SessionManager(
            redis.sub_manager('session'),
            max_session_length=self.max_ussd_session_length)

        self.wikipedia = WikipediaAPI(
            self.api_url, self.accept_gzip, self.user_agent)

    def teardown_application(self):
        return self.session_manager.stop()

    def make_options(self, options, prefix='', start=1):
        """
        Turn a list of results into an enumerated multiple choice list
        """
        joined = mkmenu(options, prefix,
         start)
        while len(joined) > self.max_ussd_content_length:
            if not options:
                break
            options = options[:-1]
            joined = mkmenu(options, prefix, start)

        return len(options), joined[:self.max_ussd_content_length]

    @inlineCallbacks
    def get_extract(self, title):
        key = self.wikipedia.url + ':' + title
        data = yield self.extract_redis.get(key)
        if data is None:
            extract = yield self.wikipedia.get_extract(title)
            data = json.dumps(extract.sections)
            # We do this in two steps because our redis clients disagree on
            # what SETEX should look like.
            yield self.extract_redis.set(key, data)
            yield self.extract_redis.expire(key, self.content_cache_time)
        else:
            extract = ArticleExtract(json.loads(data))
        returnValue(extract)

    @inlineCallbacks
    def consume_user_message(self, msg):
        log.msg("Received: %s" % (msg.payload,))
        user_id = msg.user()
        session = yield self.session_manager.load_session(user_id)
        if (not session) or (msg['content'] is None):
            session = yield self.session_manager.create_session(user_id)
            session['state'] = 'new'

        pfunc = getattr(self, 'process_message_%s' % (session['state'],))
        try:
            session = yield pfunc(msg, session)
            if session['state'] is not None:
                yield self.session_manager.save_session(user_id, session)
            else:
                yield self.session_manager.clear_session(user_id)
        except:
            log.err()
            self.reply_to(
                msg, 'Sorry, there was an error processing your request. '
                'Please try again later.', False)
            yield self.session_manager.clear_session(user_id)

    def process_message_new(self, msg, session):
        self.reply_to(
            msg, "What would you like to search Wikipedia for?", True)
        session['state'] = 'searching'
        return session

    @inlineCallbacks
    def process_message_searching(self, msg, session):
        query = msg['content'].strip()

        results = yield self.wikipedia.search(query)
        if results:
            count, msgcontent = self.make_options(results)
            session['results'] = json.dumps(results[:count])
            self.reply_to(msg, msgcontent, True)
            session['state'] = 'sections'
        else:
            self.reply_to(
                msg, 'Sorry, no Wikipedia results for %s' % query, False)
            session['state'] = None
        returnValue(session)

    def select_option(self, options, msg):
        response = msg['content'].strip()

        if response.isdigit():
            try:
                return options[int(response) - 1]
            except (KeyError, IndexError):
                pass
        self.reply_to(msg,
                      'Sorry, invalid selection. Please restart and try again',
                      False)
        return None

    @inlineCallbacks
    def process_message_sections(self, msg, session):
        selection = self.select_option(json.loads(session['results']), msg)
        if not selection:
            session['state'] = None
            returnValue(session)

        session['page'] = json.dumps(selection)
        extract = yield self.get_extract(selection)
        results = extract.get_section_titles()  # TODO:
        results = [selection] + results
        count, msgcontent = self.make_options([r for r in results])
        session['results'] = json.dumps(results[:count])
        self.reply_to(msg, msgcontent, True)
        session['state'] = 'content'
        returnValue(session)

    @inlineCallbacks
    def process_message_content(self, msg, session):
        sections = json.loads(session['results'])
        selection = self.select_option(sections, msg)
        if not selection:
            session['state'] = None
            returnValue(session)
        page = json.loads(session['page'])
        extract = yield self.get_extract(page)
        content = extract.sections[int(msg['content'].strip()) - 1]['text']
        ussd_cont = truncate_sms_with_postfix(
            content, '\n(Full content sent by SMS.)',
            self.max_ussd_content_length, self.max_ussd_unicode_length)
        self.reply_to(msg, ussd_cont, False)
        if self.sms_transport:
            sms_content = normalize_whitespace(content)
            # TODO: Decide if we want this.
            sms_content = truncate_sms(
                sms_content,
                self.max_sms_content_length, self.max_sms_unicode_length)
            bmsg = msg.reply(sms_content)
            bmsg['transport_name'] = self.sms_transport
            if self.override_sms_address:
                bmsg['to_addr'] = self.override_sms_address
            self.transport_publisher.publish_message(
                bmsg, routing_key='%s.outbound' % (self.sms_transport,))
        session['state'] = None
        returnValue(session)
